src/polyglot-notebooks-vscode-common/src/metadataUtilities.ts

```diff
@@ -347,9 +365,54 @@ export function createDefaultNotebookDocumentMetadata(): NotebookDocumentMetadat
                 }
             ],
         }
-    };;
+    };
 }
 
 function createDefaultNotebookCellMetadata(): NotebookCellMetadata {
     return {};
+}
+
+export function areEquivalentObjects(object1: { [key: string]: any }, object2: { [key: string]: any }): boolean {
+    const isObject = (object: any) => {
+        return object !== null && typeof object === 'object';
+    };
+
+    const object1Keys = Object.keys(object1);
+    const object2Keys = Object.keys(object2);
+
+    if (object1Keys.length !== object2Keys.length) {
+        return false;
+    }
+
+    for (const key of object1Keys) {
+        key;//?
+        const value1 = object1[key];//?
+        const value2 = object2[key];//?
+        const bothAreObjects = isObject(value1) && isObject(value2); //?
+        const bothAreArrays = Array.isArray(value1) && Array.isArray(value2);
+
+        if (bothAreArrays) {
+            if (value1.length !== value2.length) {//?
+                return false;
+            }
+            for (let index = 0; index < value1.length; index++) {
+                const element1 = value1[index];//?
+                const element2 = value2[index];//?
+                if (!areEquivalentObjects(element1, element2)) {
+                    return false;
+                }
+            }
+        } else if (bothAreObjects) {
+            const equivalent = areEquivalentObjects(value1, value2);
+            if (!equivalent) {
+                return false;
+            }
+        } else if (value1 !== value2) //?
+        {
+            value1;//?
+            value2;//?
```

Comment:
I am guessing the //? comments are tracking nullability? Are the above two lines needed?

Reply:
those are actually something cooler than that :) let me know if you want a demo :) :)

Reply:
Ah would love a demo :)